### PR TITLE
Feature/ted/dat 2878/datagroup class list schema

### PR DESCRIPTION
**Pull request by ted-brundage-am**

### CODE REVIEW

**Comment by eugsokolov** [comment]
retest this please

File: datamonster_api/lib/data_group.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        self.name = name
+        self.columns = columns
+        self.dm = dm
+
+    def __hash__(self):
+        return hash(self.id)
+
+    def __eq__(self, obj):
+        return isinstance(obj, DataGroup) and self.id == obj.id
+
```
Comment (eugsokolov) [inline datamonster_api/lib/data_group.py]: i noticed you have a repr for the other one, does this one not warrant one?

**Comment by eugsokolov** [comment]
retest this please

**Comment by eugsokolov** [comment]
retest this please

File: datamonster_api/lib/datamonster.py [hunk 2]
```diff
@@ -433,6 +438,69 @@ def to_json_checked(filters):
                 "Problem with filters when getting dimensions: {}".format(e)
             )
 
+    ##############################################
```
Comment (eugsokolov) [inline datamonster_api/lib/datamonster.py]: worth creating a new class? no strong feelings here

File: datamonster_api/lib/data_group.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+from .base import BaseClass
+from .errors import DataMonsterError
+import numpy as np
+
+date_regex = r'\d{4}-\d{2}-\d{2}'
```
[the remark this answers is not in the document]
Comment (ted-brundage-am) [reply datamonster_api/lib/data_group.py]: I'd say no -- if we want to make sure they're real dates, the best way is to try to instantiate date objects. If we check 1-12, then we should check 1-31 for the days, and etc.etc. I'd vote either the string format, or actually trying to instantiate dates.

File: datamonster_api/lib/data_group.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        elif np.issubdtype(column, np.number):
+            return 'number'
+        elif np.issubdtype(column, np.object_) or np.issubdtype(column, np.str_):
+            return 'string'
+
+    @staticmethod
+    def _construct_error_message(missing, extras, bad_dates):
+        msg = 'Invalid DataFrame Schema:\n'
+        if missing:
+            msg = msg + '  DataGroup could not find the following column{}:\n'.format(
```
[the remark this answers is not in the document]
Comment (ted-brundage-am) [reply datamonster_api/lib/data_group.py]: oh nice

File: datamonster_api/lib/data_group.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        if self.name in df.columns:
+            column = df[self.name]
+            if self.type_ == 'string':
+                # pandas maps strings to objects
+                return np.issubdtype(column, np.object_) or np.issubdtype(column, np.str_)
+            elif self.type_ == 'number':
+                return np.issubdtype(column, np.number)
+            elif self.type_ == 'date':
+                return column.str.match(date_regex).any()
```
[the remark this answers is not in the document]
Comment (ted-brundage-am) [reply datamonster_api/lib/data_group.py]: yeah -- good idea

File: datamonster_api/lib/data_group.py (hunk 1 above)
Comment (ted-brundage-am) [reply datamonster_api/lib/data_group.py]: The base class that DataGroup extends has a repr

File: datamonster_api/lib/datamonster.py (hunk 2 above)
Comment (ted-brundage-am) [reply datamonster_api/lib/datamonster.py]: I'd say no -- we already have methods for companies and datasources in the same class

File: datamonster_api/lib/data_group.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        return msg
+
+    def _validate_schema(self, df):
+        """Check if the schema of a provided pandas dataframe matches the expected columns"""
+        missing = []
+        extra = []
+        bad_dates = []
+
+        # Find missing columns
+        for col in self.columns:
```
[the remark this answers is not in the document]
Comment (ted-brundage-am) [reply datamonster_api/lib/data_group.py]: ugh. yes

**Comment by jlindenger** [comment]
We should test this branch against staging before merging it.

**Pull request merged**